System/Random.hs

```diff
@@ -257,6 +193,95 @@ mutableByteArrayContentsCompat :: MutableByteArray s -> Ptr Word8
 -- buildWord64 w0 w1 = ((fromIntegral w1) `shiftL` 32) .|. (fromIntegral w0)
 -- :}
 
+-- $randomgen
+--
+-- [/Example for RNG Implementors:/]
+--
+-- Suppose you want to use a [permuted congruential
+-- generator](https://en.wikipedia.org/wiki/Permuted_congruential_generator)
+-- as the source of entropy. You can make it an instance of
+-- `RandomGen`:
+--
+-- >>> data PCGen = PCGen !Word64 !Word64
+--
+-- >>> :{
+-- let stepGen :: PCGen -> (Word32, PCGen)
+--     stepGen (PCGen state inc) = let
+--       newState = state * 6364136223846793005 + (inc .|. 1)
+--       xorShifted = fromIntegral (((state `shiftR` 18) `xor` state) `shiftR` 27) :: Word32
+--       rot = fromIntegral (state `shiftR` 59) :: Word32
+--       out = (xorShifted `shiftR` (fromIntegral rot)) .|. (xorShifted `shiftL` fromIntegral ((-rot) .&. 31))
+--       in (out, PCGen newState inc)
+-- :}
+--
+-- >>> fst $ stepGen $ snd $ stepGen (PCGen 17 29)
+-- 3288430965
+--
+-- >>> :{
+-- instance RandomGen PCGen where
+--   next g = (fromIntegral y, h)
+--     where
+--       (y, h) = stepGen g
+--   split _ = error "This PRNG is not splittable"
+-- :}
+--
+-- Importantly, this implementation will not be as efficient as it
+-- could be because the random values are converted to 'Integer' and
+-- then to desired type.
+--
+-- Instead we should define (where e.g. @unBuildWord32 :: Word32 ->
+-- (Word16, Word16)@ is a function to pull apart a 'Word32' into a
+-- pair of 'Word16'):
+--
+-- >>> data PCGen' = PCGen' !Word64 !Word64
+--
+-- >>> :{
+-- instance RandomGen PCGen' where
+--   genWord8  (PCGen' s i) = (z, PCGen' s' i')
+--     where
+--       (x, PCGen s' i') = stepGen (PCGen s i)
+--       y = fst $ unBuildWord32 x
+--       z = fst $ unBuildWord16 y
+--   genWord16 (PCGen' s i) = (y, PCGen' s' i')
+--     where
+--       (x, PCGen s' i') = stepGen (PCGen s i)
+--       y = fst $ unBuildWord32 x
+--   genWord32 (PCGen' s i) = (x, PCGen' s' i')
+--     where
+--       (x, PCGen s' i') = stepGen (PCGen s i)
+--   genWord64 (PCGen' s i) = (undefined, PCGen' s i)
+--     where
+--       (x, g)           = stepGen (PCGen s i)
+--       (y, PCGen s' i') = stepGen g
+--   split _ = error "This PRNG is not splittable"
+-- :}
+--
+-- [/Example for RNG Users:/]
+--
+-- >>> setStdGen (mkStdGen 1729) >> replicateM 10 (randomRIO (1,6)) :: IO [Word32]
+-- [2,4,5,4,3,2,4,2,4,3]
+--
+-- Suppose you want to simulate rolls from a dice (yes I know it's a
+-- plural form but it's now common to use it as a singular form):
+--
+-- >>> :{
+-- let randomListM :: (MonadRandom g m, Num a, Uniform a) => g -> Int -> m [a]
+--     randomListM gen n = replicateM n (uniform gen)
+-- :}
+--
+-- >>> :{
+-- let rolls :: [Word32]
+--     rolls = runGenState_
+--               (PCGen 17 29)
+--               (randomListM PureGenI 10 >>= \xs -> return $ map ((+1) . (`mod` 6)) xs)
```

Comment:
```suggestion
--               (randomListM 10 (1, 6) PureGenI)
```
(From https://github.com/idontgetoutmuch/random/pull/43/files#diff-4a522d098452ebfc9be2d6291588f5bcR106-R114)

I think this is an improvement because the `mod` method could be biased, which would not be demonstrating best practice.